File: backend/app/schemas/prediction.py
"""`GET /api/v1/batteries/{battery_id}/prediction`'s response shape
(Roadmap 4.5, Contract section 47).

The Contract's own example only shows the *successful* case
(`prediction_method="ml"`, a real number, a real timestamp). This adds
`prediction_available` and `reason` alongside it -- the same two fields
`BaselinePredictionResult` (Roadmap 4.1) already uses -- so the
"unavailable" case the Contract also requires ("returns a defined reason
rather than a fabricated numeric prediction") has an actual field to put
that reason in, instead of overloading `prediction_method` to mean two
different things.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class PredictionResponse(BaseModel):
    """A successful prediction has `prediction_available=True`,
    `reason=None`, and every other field populated. An unavailable one has
    `prediction_available=False`, a non-null `reason`, and
    `predicted_minutes_to_critical` / `predicted_critical_timestamp` /
    `prediction_method` / `model_version` all `None` -- never a fabricated
    number standing in for "we don't actually know."""

    battery_id: str
    current_soc: float | None
    critical_soc: float
    prediction_available: bool
    predicted_minutes_to_critical: float | None
    predicted_critical_timestamp: datetime | None
    prediction_method: str | None
    model_version: str | None
    reason: str | None

File: frontend/src/types/prediction.ts
// Mirrors backend/app/schemas/prediction.py's PredictionResponse exactly
// (Roadmap 4.5, Contract section 47). Same by-hand-sync tradeoff as the
// other types/*.ts files -- no codegen yet.
//
// `prediction_available=false` always comes with a `reason` and every other
// prediction field `null` -- there's no partial prediction (see
// prediction_service.py's BaselinePredictionResult docstring, which this
// mirrors). `predicted_critical_timestamp` stays `string`, not `Date`, for
// the same reason as battery.ts's `last_seen` -- fetch() never parses dates
// automatically.

export interface PredictionResponse {
  battery_id: string
  current_soc: number | null
  critical_soc: number
  prediction_available: boolean
  predicted_minutes_to_critical: number | null
  predicted_critical_timestamp: string | null
  prediction_method: string | null
  model_version: string | null
  reason: string | null
}
